### PR TITLE
test: Phase 4 -- real end-to-end identity-taint metamorphic coverage

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Phase 4 of [`docs/contribute/plans/bug-class-regression-testing.md`](https://github.com/abicheck/abicheck/blob/main/docs/contribute/plans/bug-class-regression-testing.md), generalizing the `identity.environment_taint` bug class (#837 → #843 → #846 → #868).

## Motivation

The plan's Phase 4 invariant: *"canonical identity (finding IDs, type/function identity keys, node IDs) is a function of semantic scope and source identity — never of checkout root, absolute path spelling, temp-directory location, path separator, or unrelated line/column drift from an edit elsewhere in the file."*

The two existing seed tests (`test_castxml_anonymous_type_location.py`, `test_anon_type_location_properties.py`) cover `strip_anonymous_type_location` thoroughly at the unit level, but — per this bug class's own `tests/regressions/manifest.py` known-gap entry — both feed a hand-built AST-node/XML fragment into an internal parser class directly: no real compiler subprocess, no CLI, no python-api call. This PR closes that gap with real, end-to-end coverage.

## Changes

- **`tests/test_identity_taint_end_to_end.py`** (new): real g++/direct-clang subprocess-based end-to-end coverage through `abicheck.dumper.dump()`/`abicheck.checker.compare()`:
  - Relocating the checkout root, a symlinked checkout root, unrelated blank-line/comment insertion (line drift), and reordering unrelated declarations — each asserted `Verdict.NO_CHANGE` with zero emitted changes. Verified non-vacuous: `source_header` genuinely differs between the two checkouts while `compare()` still reports no findings.
  - One `@pytest.mark.integration` cross-backend variant of the core #843 checkout-relocation case, verified against the real castxml backend too (matching `test_clang_castxml_origin_parity.py`'s marker discipline from the Phase 3 PR).
  - Negative control (do not over-merge), reusing AGENTS.md's "using-declaration" known-gap entry's own counterexamples: two lambdas in one header, and two same-named nested records in different namespaces, both confirmed to stay genuinely distinct identities across the identical relocation transformation that collapses every positive case above to `NO_CHANGE`.
  - Path-separator and archive-member-order transformations from the plan's own list are deliberately not attempted (no Windows-style filesystem or static-archive concept applies to this sandbox/class), and a compilation-database-root change is left to the pre-existing L3-focused suites that already cover it — both recorded honestly in the manifest rather than faked.
- **`tests/test_lambda_identity_ordinal.py`**: adds `TestL5SourceGraphIdentitiesAreNotRenumbered`, a dedicated canary for the L5 residual AGENTS.md already documents ("the L5 source graph's own node identities are not renumbered alongside the flat snapshot's closure markers"). Asserts the residual's own bound per `KnownGap`'s docstring rule: a closure marker in `build_source.source_graph` is left untouched by `renumber_anonymous_closure_identities` while the equivalent flat-snapshot field IS renumbered, in one snapshot — a future fix in either direction now fails this test loudly instead of the gap silently closing or widening unnoticed.
- **`tests/regressions/manifest.py`**: registers both new seed tests under `identity.environment_taint`, wires the L5 gap to its new `canary_test`, and rewrites the cross-surface/cross-backend known-gap text to describe what this phase closed and what remains honestly deferred.

Test-only change; no `abicheck/**/*.py` production code is touched, so no changelog fragment is needed.

## Checklist

- [x] Tests added / updated for new behaviour
- [ ] Changelog fragment added (`scriv create`) if this touches `abicheck/**/*.py` — see `changelog.d/README.md` (N/A — test-only change)
- [ ] Docs updated if user-visible behaviour changed (N/A — no behavior change)
- [x] `pre-commit run --all-files` passes locally (ruff check/format clean on touched files; `tests/regressions/manifest.py` integrity tests pass)
- [x] No new `mypy` or `ruff` errors introduced

---
_Generated by [Claude Code](https://claude.ai/code/session_018USuGdPgyVaPK7z6iB2cKd)_

### CODE REVIEW

File: tests/test_lambda_identity_ordinal.py [hunk 1]
```diff
@@ -905,3 +908,53 @@ def test_fact_producer_resolves_after_renumbering(self) -> None:
...
+        node_label = "(lambda at /src/x/widget.h:522:26)"
+        snap = AbiSnapshot(
+            library="lib.so",
+            version="1",
+            types=[_record(owner, qualified=owner)],
+            build_source=BuildSourcePack(
+                root=Path("/src/x"),
+                source_graph=SourceGraphSummary(
+                    nodes=[GraphNode(id="n1", kind="closure", label=node_label)]
+                ),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_lambda_identity_ordinal.py:959]: Fixed in 224e90e: the canary now constructs a production-shaped `decl://`-prefixed node id (matching `graph_facts._decl_node_id`'s real scheme) plus a real edge referencing it, and asserts the id/edge-endpoint stay unchanged across the renumbering pass, not just the label.

One thing worth recording: while implementing this I found `SourceGraphSummary.__post_init__` (via `add_node`/`ensure_facts_and_resolve`) *already* normalizes a decl/type node's id and label at construction time — but that's the pre-existing, unrelated checkout-path-taint normalization (`_normalize_graph_identity`), not this fix's ordinal renumbering, and it leaves the raw `:line:col` intact. The test now captures the post-construction (pre-renumber) baseline rather than asserting against the literal constructor arguments, so it isn't tripped up by that unrelated mechanism.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+
+def _dump(so: Path, header: Path, header_backend: str = "clang"):
+    return dump(so, [header], header_backend=header_backend, public_headers=[header])
+
+
+_HEADER = """
+#pragma once
+namespace lib {
+struct Point { int x; int y; };
+int add(int a, int b) noexcept;
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:128]: Fixed in 224e90e: `_HEADER`/`_SOURCE` now include a `template <class F> inline int invoke_with(F f)` instantiated over a lambda alongside the ordinary declarations, and the noisy/reordered header variants place it so its line:col genuinely shifts (below the inserted comments in the drift case, moved to the front in the reorder case). Every positive test now asserts (`_assert_exercises_closure_taint`) the closure-tainted symbol is actually present before trusting the NO_CHANGE oracle.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        """Cross-backend closure of this same case: the #843 bug was
+        originally a CastXML-specific ctor/dtor identity taint, so the
+        checkout-relocation invariant is verified against the real castxml
+        backend here too, not only direct-clang above -- marked
+        ``integration`` since it needs castxml in addition to clang/g++
+        (see ``test_clang_castxml_origin_parity.py`` for the identical
+        marker-discipline reasoning)."""
+        if not _have("castxml"):
+            pytest.skip("castxml required for this cross-backend variant")
+        _require_toolchain()
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 224e90e: the castxml checkout-relocation variant no longer calls `_require_toolchain()` (which required clang unconditionally). It now checks only `g++` and `castxml` — the two tools it actually invokes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+import pytest
+
+from abicheck.checker import Verdict, compare
+from abicheck.dumper import dump
+
+pytestmark = pytest.mark.skipif(
+    not sys.platform.startswith("linux"),
+    reason="real g++/clang ELF end-to-end test is Linux-scoped",
+)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:99]: This is a deliberate, pre-existing repo convention rather than an oversight — I'm not changing it. `tests/test_clang_header_backend_integration.py` (already merged, predates this PR) establishes exactly this pattern: a module that needs only clang+g++ stays unmarked and gated purely by a `_have()`-based skip, precisely so it *does* run in the fast lane on a host that has those tools — see that file's own comment at line 53: "the module isn't marked `integration` (that gate requires castxml)... a per-test `@pytest.mark.integration` so a castxml-having host's fast lane skips it." This module's docstring cites that exact precedent (line 62-64) and follows it identically: unmarked for the clang+g++-only tests, `@pytest.mark.integration` only on the one test that additionally needs castxml.

Marking the whole module `integration` would mean these tests never run at all in CI's default fast/PR lane (which excludes `integration`), which is a bigger loss than the runtime-variance concern — the whole point of Phase 4 was closing the "no real compiler subprocess" gap the seed tests had. Happy to revisit if there's a reason this precedent itself should change, but that would apply to `test_clang_header_backend_integration.py` too, not just this new module.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+def _assert_exercises_closure_taint(snap: object) -> None:
+    """The positive (NO_CHANGE-expected) cases in this module are only
+    meaningful if the fixture actually produces a closure/lambda-embedded
+    identity -- otherwise a regression in path/line-drift handling for
+    exactly that identity shape (the historical taint mechanism) would go
+    uncaught while the test stays green for an unrelated reason (ordinary
+    named declarations don't embed source location in their identity at
+    all). Fails loudly, not silently, if the fixture stops producing one."""
+    names = {f.mangled for f in snap.functions}  # type: ignore[attr-defined]
+    assert any("invoke_with" in n for n in names), (
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 9910744: `_assert_exercises_closure_taint` now checks for a `(lambda` marker in a function's parameter/return type spelling, not a mangled-name substring — confirmed the generic template pattern really is emitted with bare `mangled == "invoke_with"`, so the old check would have stayed green even with the specialization gone.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+        so, header = _build(root, header_text, source_text)
+        snap = _dump(so, header)
+
+        # The two lambda-parameterized instantiations of call_with must be
+        # genuinely distinct types/functions within the SAME snapshot --
+        # not merely "not equal to each other by accident", but present as
+        # two separate entries.
+        lambda_functions = [f for f in snap.functions if "call_with" in f.mangled]
+        distinct_mangled = {f.mangled for f in lambda_functions}
+        assert len(distinct_mangled) >= 2, (
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 9910744: the negative control now asserts distinct CANONICAL (post-renumbering) parameter-type spellings — `(lambda:api.h#N)` — filtered to entries actually containing `(lambda`, rather than raw mangled linker symbols. You're right that the old check was vacuous: `run_one`/`run_two` are different enclosing functions, so real Itanium mangling already guarantees distinct symbols regardless of any identity-collapse bug.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+        so_b, header_b = _build(root_b, _HEADER, _SOURCE)
+
+        snap_a = _dump(so_a, header_a)
+        snap_b = _dump(so_b, header_b)
+
+        _assert_exercises_closure_taint(snap_a)
+        _assert_exercises_closure_taint(snap_b)
+        result = compare(snap_a, snap_b)
+        assert result.verdict is Verdict.NO_CHANGE
+        assert result.changes == []
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:241]: Fixed in 9910744: added `TestFindingIdentityIsCheckoutPathInvariant`, which introduces one real, controlled ABI addition (a second lambda-parameterized `invoke_with<...>` instantiation) and asserts `report_canonical_finding_id` produces the same id set under two different checkout roots.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+int add(int a, int b) noexcept { return a + b; }
+int Widget::value() const { return hidden_; }
+int touch_lambda() { return uses_lambda(); }
+}
+"""
+
+
+def _require_toolchain() -> None:
+    if not (_have("clang") and _have("g++")):
+        pytest.skip("clang and g++ are required for this end-to-end test")
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 9910744: `_require_toolchain` now checks `clang++` instead of the bare `clang` driver, matching what `dump()`'s default `compiler="c++"` actually resolves to.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:241]: Good catch — confirmed empirically: `FUNC_ADDED`'s mangled symbol and `UNNAMED_TYPE_IN_PUBLIC_ABI`'s `{lambda()#N}` description discriminator are both the real compiler's own structural ordinal, entirely unrelated to this repo's `renumber_anonymous_closure_identities`, so neither actually depends on this fix at all.

Fixed in 25d2228 with a new `test_type_bearing_finding_gets_the_same_canonical_id_under_different_roots`. The direct-clang backend doesn't emit closures as standalone `RecordType`s (confirmed empirically — only `Function.params[*].type` carries the `(lambda:...)` spelling), and castxml isn't available in this sandbox to produce a genuine type-level finding through `compare()` itself, so this builds the `Change` directly — but from a real, `dump()`-produced parameter-type spelling (not a hand-built AST fragment), under two different checkout roots. Since `renumber_anonymous_closure_identities` already runs inside `dump()` before any `Change` is constructed, this is exactly the value a real finding would carry in production, and `report_canonical_finding_id`'s NORMALIZED tier folds it in verbatim for this kind — so this does exercise the actual risk.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 25d2228: the canary now re-fetches `snap.build_source.source_graph` after calling `renumber_anonymous_closure_identities`, rather than reusing the `graph` reference captured beforehand — a future fix that reassigns a rewritten object onto `snap.build_source` instead of mutating in place would otherwise leave the stale reference showing pre-fix values forever.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+        result_a = compare(old_a, new_a)
+        result_b = compare(old_b, new_b)
+
+        from abicheck.finding_identity import report_canonical_finding_id
+
+        def _added_lambda_canonical_ids(changes: object) -> set[str]:
+            candidates = [
+                c
+                for c in changes  # type: ignore[attr-defined]
+                if c.symbol and "uses_lambda2" in c.symbol
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:922]: placeholder

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:922]: (Sorry, that was a stray placeholder reply meant for a different thread — please ignore. Real reply below.)

Fixed in 33dd7dd: this test now builds root B from a semantics-preserving header variant with unrelated comments inserted before the lambda, rather than compiling `_HEADER` at the identical line under both roots. Verified by manually disabling `renumber_anonymous_closure_identities`: the two raw spellings then genuinely differ (`(lambda:api.h:13:47)` vs `(lambda:api.h:17:47)`), confirming the test now correctly fails if ordinal renumbering breaks.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/regressions/manifest.py [hunk 10]
```diff
@@ -263,25 +263,61 @@ class BugClass:
         seed_tests=(
             "tests/test_castxml_anonymous_type_location.py",
             "tests/test_anon_type_location_properties.py",
+            "tests/test_lambda_identity_ordinal.py",
+            "tests/test_identity_taint_end_to_end.py",
         ),
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/regressions/manifest.py:268]: Fixed in 33dd7dd: `identity.environment_taint`'s `BugClass` now sets `axes={"frontend": ("clang", "castxml")}`, matching the seed tests that genuinely invoke both real backends.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+                        return p.type
+            raise AssertionError(
+                "expected a lambda-parameterized invoke_with<...> "
+                "instantiation in the snapshot"
+            )
+
+        spelling_a = _closure_param_spelling(tmp_path / "checkout_a")
+        spelling_b = _closure_param_spelling(
+            tmp_path / "an/unrelated/deeper/checkout_b"
+        )
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py]: Fixed in 33dd7dd: root B now builds from a semantics-preserving header variant with unrelated comments inserted before the lambda (shifting its line), rather than compiling `_HEADER` at the identical line under both roots. Verified by manually disabling `renumber_anonymous_closure_identities`: the two raw spellings then genuinely differ (`(lambda:api.h:13:47)` vs `(lambda:api.h:17:47)`), confirming this now correctly fails if ordinal renumbering breaks.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_lambda_identity_ordinal.py [hunk 12]
```diff
@@ -905,3 +908,101 @@ def test_fact_producer_resolves_after_renumbering(self) -> None:
...
+        # identities" baseline, rather than asserting against the literal
+        # strings passed into the constructor.
+        graph = snap.build_source.source_graph  # type: ignore[union-attr]
+        assert graph is not None
+        pre_renumber_id = graph.nodes[0].id
+        pre_renumber_label = graph.nodes[0].label
+        pre_renumber_edge_src = graph.edges[0].src
+        assert "522" in pre_renumber_label and "26" in pre_renumber_label
+
+        renumber_anonymous_closure_identities(snap)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_lambda_identity_ordinal.py:980]: I checked the actual production call trace before responding, and the claimed ordering doesn't hold for the paths that matter most here.

For ELF/PE/Mach-O (`service.py`'s `_run_dump_uncached`, reached by every ordinary `dump()` call — `cli_dump_helpers.perform_elf_dump`'s ELF path included), the real order is the **opposite** of what's described: `_attach_header_graph` runs first (assigning a brand-new `BuildSourcePack`/`SourceGraphSummary` onto `snap.build_source` — confirmed by reading `service_header_graph_attach.py:326`, `snap.build_source = pack`, a wholesale reassignment, not an in-place mutation), and `renumber_anonymous_closure_identities(snap)` is called *last*, after that assignment. `_finish_native_snapshot`'s own docstring states this explicitly: "renumbered here exactly once, after `attach_clang_layout`". So for these paths, my canary's construction order — pre-populate `build_source`, then call `renumber_anonymous_closure_identities` — is exactly the real production sequence, not a simplification of it.

The `hybrid` backend (castxml+clang merge) does reverse the order the way you describe (`service.py` around line 790: renumber-then-attach). But note the *observable* bug is identical either way: whichever step runs second, the L5 graph reaching the caller was never touched by the ordinal-renumbering logic (in the hybrid case, `_attach_header_graph` runs *after* renumbering already finished, building a fresh graph from a fresh reparse that renumbering never sees). So the residual this canary pins — "the L5 side is untouched by this fix" — holds for every real dump path, just via two different call orderings, and my canary reproduces the dominant (ELF/PE/Mach-O) one verbatim.

I don't think a fix is more likely to land inside `_attach_header_graph`'s own construction than inside the one function every path already funnels through at some point (`renumber_anonymous_closure_identities`) — especially since the two paths disagree on ordering, so a fix living in the attach function alone wouldn't even see a consistent input across backends. A genuinely production-shaped version of this canary (invoking `_attach_header_graph` for real) would need a real header + clang subprocess, which is exactly the "hand-built fragment vs. real subprocess" distinction this whole Phase 4 effort is about — that level of fidelity belongs in `test_identity_taint_end_to_end.py`, not this synthetic unit-level file, and I'd rather leave it as a documented possible follow-up than force it into this canary under continued review pressure.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
`integration-tests (ubuntu-24.04)` failed on 33dd7dd: `test_two_independent_checkouts_compare_identical_via_castxml` — real castxml resolves the closure argument of a bare function-template parameter (`invoke_with<F>`) to the unresolvable-type sentinel `"?"`, not to `"(lambda:...)"` the way direct-clang does, so `_assert_exercises_closure_taint`'s marker search correctly found nothing and failed rather than passing vacuously.

Fixed in 33ec67f by adding a class template instantiated over a closure (`Guard<F>`) alongside the existing function-template one — castxml's own documented closure-taint mechanism (AGENTS.md's "Lambda-closure churn" entry) is specifically synthetic ctor/dtor keys on a class template instantiated with a lambda, which is the shape the fixture was missing. The taint-marker check now also scans `Function.mangled` and `RecordType.name`/`qualified_name`, not just parameter/return type spellings, so whichever field a given backend embeds the marker in is found.

Verified locally against the real direct-clang backend; castxml isn't available in this sandbox to reproduce the exact CI failure against directly, but the added construct is the exact mechanism AGENTS.md documents as castxml's real, verified behavior. Watching for the next CI run to confirm.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 13]
```diff
@@ -0,0 +1,719 @@
...
+        real_root.mkdir()
+        so, header = _build(real_root, _HEADER, _SOURCE)
+
+        symlink_root = tmp_path / "checkout_via_symlink"
+        symlink_root.symlink_to(real_root, target_is_directory=True)
+        so_via_link = symlink_root / so.name
+        header_via_link = symlink_root / header.name
+
+        snap_direct = _dump(so, header)
+        snap_via_symlink = _dump(so_via_link, header_via_link)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:308]: Confirmed exactly what you describe empirically (real g++/clang dump): `source_location`/`source_header` for both invocations already record the identical resolved REAL path — clang canonicalizes the symlink itself before abicheck's model sees it. I also checked whether the closure marker could show a difference even with renumbering disabled and it can't: the marker's spelling never carries a directory component at all (just the bare filename + line:col), so there's no lexical symlink spelling to preserve there either way.

Given that, your literal suggestion ("assert the raw producer paths differ") would be a false assertion against real clang — there's no abicheck-side symlink-canonicalization logic to test here, since clang already does it upstream. Fixed in bc43aed by rewriting the test's docstring to state this plainly (rather than the previous speculative "unless the extractor canonicalizes it" hedge) and adding a real assertion that *proves* the canonicalization claim (both sides record the real, non-symlinked path) instead of leaving it implicit — making explicit that this test's actual invariant is a narrower end-to-end sanity check, not a test of abicheck-specific logic.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 14]
```diff
@@ -0,0 +1,742 @@
...
+        )
+
+        # Now relocate the whole checkout and confirm the SAME two
+        # instantiations are still distinct from each other post-move
+        # (not merged into one by the relocation), while the snapshot as a
+        # whole still compares NO_CHANGE against its own un-relocated self.
+        relocated_root = tmp_path / "relocated" / "checkout"
+        relocated_root.mkdir(parents=True)
+        so2, header2 = _build(relocated_root, header_text, source_text)
+        snap2 = _dump(so2, header2)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:515]: Fixed in a8ed970: added line-drift and reordering variants for both negative-control fixtures.

Worth flagging what this turned up: the two-lambda REORDERING variant genuinely fails a naive NO_CHANGE expectation — swapping `run_one`/`run_two`'s declaration order swaps which ordinal each lambda gets, and since `compare()` matches functions by their real, order-independent mangled symbol, the matched pair's ordinal-derived parameter spelling differs, producing a real finding. This isn't a bug — `qualified_name_segments.py`'s own module docstring already documents this exact scope boundary ("As long as an edit doesn't reorder... same-header, same-kind lambdas relative to each other, both sides... assign the identical ordinal"). So that one test now asserts the actual negative-control invariant (the two closures never collapse into ONE shared identity within either snapshot) plus the honest, real finding `compare()` reports for reordering, rather than a false NO_CHANGE expectation. The line-drift variants (which don't change relative order) do assert NO_CHANGE, as expected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 15]
```diff
@@ -0,0 +1,970 @@
...
+inline int run_one() { return call_with([]() { return 1; }); }
+inline int run_two() { return call_with([]() { return 2; }); }
+}
+"""
+        reordered_header = """
+#pragma once
+namespace lib {
+template <class F> int call_with(F f) { return f(); }
+inline int run_two() { return call_with([]() { return 2; }); }
+inline int run_one() { return call_with([]() { return 1; }); }
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:677]: Fair — the previous round's test swapped run_one/run_two themselves (the already-documented same-kind-reorder limitation), not the unrelated-declaration-reordering transformation you originally asked for. Added `test_two_lambdas_stay_distinct_when_an_unrelated_declaration_moves` in e2d990e: keeps run_one/run_two in their original relative order and moves a third, unrelated declaration (`unrelated_fn`) around them instead, requiring both distinct identities and `NO_CHANGE` — the shape that actually exercises the untested case without touching the documented ordinal-swap boundary. The previous test (swapping the two lambdas directly) is kept as-is, since it's a real, distinct, correctly-documented case in its own right.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_identity_taint_end_to_end.py [hunk 16]
```diff
@@ -0,0 +1,1033 @@
...
+        snap_b = _dump(so_b, header_b)
+
+        assert len(_lambda_param_spellings(snap_a.functions)) >= 2
+        assert len(_lambda_param_spellings(snap_b.functions)) >= 2
+
+        result = compare(snap_a, snap_b)
+        assert result.verdict is Verdict.NO_CHANGE
+        assert result.changes == []
+
+    def test_two_lambdas_stay_distinct_after_reordering(self, tmp_path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_identity_taint_end_to_end.py:643]: Good catch — confirmed: it was dropped when I rewrote this test's docstring in the previous round. Fixed in 2dd7f44. Also swept every other test method in the file to confirm none has the same gap.

---
_Generated by [Claude Code](https://claude.ai/code)_
